### DISCUSSION TITLE
How to execute with a command and avoid configuration

### DISCUSSION THREAD

**Discussion by robercoding** [asarkar/build-time-tracker#48] [Q&A] [answered]
First of all I'm not a gradle expert, I'm still learning and find myself in the first steps to understand gradle lifecycle and how to improve build times
What I've noticed is that this plugin is being executed in every build, and I personally don't want to know how much time it takes in every Run. So my thoughts after seeing this happen is that I'd like to register a task and avoid configuration phase and not execute it in every build

I was looking forward to run in a similar way to other plugins ./gradlew :app:trackTime and not execute the plugin in every run

But maybe this is not possible, and I'd like to know what other options are there and if I'm missing some knowledge then feel free to point it out. As I said not an expert but looking forward to understand what can be done/improved

Thank you!

**Comment by asarkar** [comment]
> Plugin is executed every build

A plugin is not executed, tasks are executed. The plugin's job is to track the time taken by the tasks in the build, and that's what it does. If you don't want the plugin to do that, why would you include it in your build anyway?

> I'd like to register a task and avoid configuration phase and not execute it in every build

Sure, but what does it have to do with this plugin? The plugin doesn't register its own tasks. If you're asking about configuration caching, this plugin is compatible with Gradle configuration cache. 

> ./gradlew :app:trackTime

`:app:trackTime` is a Gradle task, not a Gradle plugin. You can skip Gradle tasks, you can't skip a plugin.

> not execute the plugin in every run

Under what condition(s) do you want the plugin to "execute", whatever execute means?

**Reply by robercoding** [reply]
Alright yeah I explained myself pretty bad and seems that I'm missing lot of knowledge

if I understand correctly a plugin contains tasks, so whenever I run the app it goes through gradle lifecycles, in the last gradle phase ("Execution") this task is being ran (And that's what it was designed for)
I want to make it optional, like when I run the `ktlintFormat` task or when I use detekt plugin using the `detektBaseline` task

To give you some context of why I'm asking this: _I want to do it this way so in local machines build time tracker tasks are not executed automatically in every run app but run it in a server every day by typing some task command or run it_. Maybe there's a better alternative?

**Comment by asarkar** [comment, marked as answer]
> a plugin contains tasks

Nope, not this one, it works by observing Gradle build lifecycle. Some plugins may register tasks, but they are _not_ required to.

> run it in a server every day

That's not the purpose of this plugin. It tracks individual builds, not run on a schedule. You'll have to roll your own mechanism to do that.

**Reply by robercoding** [reply]
> Nope, not this one. Some plugins may register tasks, but they are 

- Alright I've learned a new thing 😉 

>  It tracks individual builds, not run on a schedule. You'll have to roll your own mechanism to do that.

- I already have a Jenkins pipeline executing a build every day, so I can get the generated .csv, maybe I should instead apply the plugin there in the daily build instead of having it in the local project (_if I want to reach the main purpose of not executing it in local machine_).

Now last question, do you know if having this plugin also affects build times, and if yes, do you have an idea of how much it affects or how I could also track that? Let's say I have over 100 modules and build times are over 20 min, is there a chance of having the BuildTimeTracker plugin also increases the build time? (_Maybe 30 seconds, 1 min_)

**Comment by asarkar** [comment]
> if having this plugin also affects build times

I've not measured it. The plugin doesn't do network calls or complicated number crunching, so, the time taken is mostly for generating the output, i.e. writing to the console or CSV, which is negligible compared to the total build time (definitely not in the ballpark of 30 seconds or 1 minute).

I hope I've answered your questions satisfactorily because I'm going to lock this question down, as this is turning into less about the plugin, more about advising on your particular project requirement.
